KAFKA-2489: add benchmark for new consumer

granders opened the pull request:
@ewencp 
The changes here are smaller than they look - mostly refactoring/cleanup.
- ConsumerPerformanceService: added new_consumer flag, and exposed more command-line settings
- benchmark.py: refactored to use `@parametrize` and `@matrix` - this reduced some amount of repeated code
- benchmark.py: added consumer performance tests with new consumer (using `@parametrize`)
- benchmark.py: added more detailed test descriptions
- performance.py: broke into separate files

Commit:
Fixed typo in ProducerPerformance.java

Commit:
Fixed ConsumerPerformance.scala with new consumer: incorrect inequality previously caused this to time out without consuming any messages with new consumer.

Commit:
Updated consumer throughput tests to run with new and old consumer. Refactored with @parametrize and @matrix to reduce duplicated code.

Commit:
Refactored performance.py - broke services into individual files. This does not change the way external users would import and use the performance services however.

granders commented:
Tests in this branch passed:
http://jenkins.confluent.io/job/kafka_system_tests_branch_builder/57/console

ewencp commented on core/src/main/scala/kafka/tools/ConsumerPerformance.scala:
```diff
@@ -100,11 +100,13 @@ object ConsumerPerformance {
+    var lastConsumedTime = System.currentTimeMillis
+
+
+    while(messagesRead < count && System.currentTimeMillis() - lastConsumedTime < timeout) {
```
I'm confused what got "fixed" here. The inequality seems the same except you changed it from `<=` to `<` and rearranged it. What exactly is the bug with using the original version with the new consumer?

granders replied:
Yes, my mistake - this is definitely (functionally) equivalent.

The thing I was attempting to fix was the `0 throughput` that you are finding when you run it. More on that below.

On the other hand, this makes the intent clearer (to me at least?).

ewencp commented on tests/kafkatest/services/performance/__init__.py:
```diff
@@ -0,0 +1,19 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from performance import PerformanceService, parse_performance_output
```
These seem like internal implementation details for this module. Any reason to expose them here? You can of course always get to them via `kafkatest.services.performance.performance`, but putting them here suggests they are part of the public interface.

granders replied:
I was thinking of projects like `muckrake` in which we have services inheriting directly from `PerformanceService`

`parse_performance_output` however is not actually used... anywhere - deleted this.

ewencp commented on tests/kafkatest/services/performance/consumer_performance.py:
```diff
@@ -0,0 +1,114 @@
+        self.settings = settings
+
+        # These less-frequently used settings can be updated manually after instantiation
+        self.fetch_size = 1024 * 1024
```
Can we not set these and have them use defaults instead? Using defaults where possible is useful for making sure we don't see regressions due to changes in those defaults. In fact, it'd be a lot better in my opinion if the ConsumerPerformance tool didn't include these defaults when there are already defaults in the underlying consumer configuration (e.g. fetch size, socket buffer size) since the current setup obscures regressions due to any changes in defaults in the consumer itself (whether accidental or not).

ewencp commented on tests/kafkatest/services/performance/producer_performance.py:
```diff
@@ -0,0 +1,68 @@
+        try:
+            self.results[idx-1] = parse_stats(last)
+        except:
+            self.logger.error("Bad last line: %s", last)
```
I know this is just a refactoring, but shouldn't this be considered a test failure and throw the exception? I have a feeling this code is a holdover from a long time ago before there was a test runner, let alone one that handled exceptions properly and converted them into test failures.

granders replied:
Good catch - fixed

ewencp commented on tests/kafkatest/services/performance/producer_performance.py:
```diff
@@ -0,0 +1,68 @@
+            }
+        last = None
+        for line in node.account.ssh_capture(cmd):
+            self.logger.debug("Producer performance %d: %s", idx, line.strip())
```
Again, I realize this was just refactored, but when going to check the logs, I found it annoying that they were interleaved in the normal test run output. Could we make producer and consumer perf services save to a log file, have them collected, and only do the collection of stats here?

ewencp commented on tests/kafkatest/tests/benchmark_test.py:
```diff
@@ -40,97 +47,71 @@ def __init__(self, test_context):
-        self.logger.info("Single producer, sync 3x replication: %s" % data)
-        return data
+    @matrix(message_size=[10, 100, 1000, 10000, 100000])
+    def test_producer_throughput_multiple_message_size(self, message_size):
```
This name could be shortened now that it's not doing multiple message sizes in each loop. The addition of the `message_size=X` component to the name should be clear even with a shorter name.

ewencp commented:
@granders Can we have that job run everything, including the `sanity_checks`? Currently the mirror maker sanity check is not passing for me (though it seems it is broken on trunk, not by this PR). I'm guessing you only run the tests directory because you're thinking of the sanity checks as useful for test developers to get quick feedback, but they are useful tests that we should always run.

Also, the tests that use new consumer are always giving me 0 for output. The logs confirm that the tool itself is indicating zeros:

```
Consumer performance 1: 2015-08-30 20:00:53:922, 2015-08-30 20:01:06:701, 0.0000, 0.0000, 0, 0.0000
```

Any idea what's going on there?

granders commented:
I also observed the `0 throughput`, and I tricked myself into thinking I had fixed it.

I added some counters in the loop to see what is happening in the consumption loop, and when I run ConsumerPerformance directly (after running the test with `--no-teardown` option),  I consistently see 11 - 13 `val records = consumer.poll(100)` before any calls to `poll` yield any records.

In other words, slightly more than 1000ms before `poll` yields non-zero results, and therefore the loop is timing out.

I had a 2 second timeout here so the loop wouldn't quit before poll received messages

```
SLF4J: Class path contains multiple SLF4J bindings.
SLF4J: Found binding in [jar:file:/vagrant/core/build/dependant-libs-2.10.5/slf4j-log4j12-1.7.6.jar!/org/slf4j/impl/StaticLoggerBinder.class]
SLF4J: Found binding in [jar:file:/vagrant/tools/build/dependant-libs-2.10.5/slf4j-log4j12-1.7.6.jar!/org/slf4j/impl/StaticLoggerBinder.class]
SLF4J: See http://www.slf4j.org/codes.html#multiple_bindings for an explanation.
SLF4J: Actual binding is of type [org.slf4j.impl.Log4jLoggerFactory]
loops: 0, zonZeroPolls: 0
messagesRead: 0
loops: 1, zonZeroPolls: 0
messagesRead: 0
loops: 2, zonZeroPolls: 0
messagesRead: 0
loops: 3, zonZeroPolls: 0
messagesRead: 0
loops: 4, zonZeroPolls: 0
messagesRead: 0
loops: 5, zonZeroPolls: 0
messagesRead: 0
loops: 6, zonZeroPolls: 0
messagesRead: 0
loops: 7, zonZeroPolls: 0
messagesRead: 0
loops: 8, zonZeroPolls: 0
messagesRead: 0
loops: 9, zonZeroPolls: 0
messagesRead: 0
loops: 10, zonZeroPolls: 0
messagesRead: 0
loops: 11, zonZeroPolls: 1
messagesRead: 49932
loops: 12, zonZeroPolls: 1
messagesRead: 49932
loops: 13, zonZeroPolls: 1
messagesRead: 49932
loops: 14, zonZeroPolls: 1
messagesRead: 49932
loops: 15, zonZeroPolls: 1
messagesRead: 49932
loops: 16, zonZeroPolls: 1
messagesRead: 49932
loops: 17, zonZeroPolls: 2
messagesRead: 99864
loops: 18, zonZeroPolls: 2
messagesRead: 99864
loops: 19, zonZeroPolls: 2
messagesRead: 99864
loops: 20, zonZeroPolls: 2
messagesRead: 99864
loops: 21, zonZeroPolls: 2
messagesRead: 99864
loops: 22, zonZeroPolls: 2
messagesRead: 99864
loops: 23, zonZeroPolls: 2
messagesRead: 99864
loops: 24, zonZeroPolls: 3
messagesRead: 133152
loops: 25, zonZeroPolls: 4
messagesRead: 149796
loops: 26, zonZeroPolls: 4
messagesRead: 149796
loops: 27, zonZeroPolls: 4
messagesRead: 149796
loops: 28, zonZeroPolls: 5
messagesRead: 166440
loops: 29, zonZeroPolls: 5
messagesRead: 166440
loops: 30, zonZeroPolls: 6
messagesRead: 183084
loops: 31, zonZeroPolls: 7
messagesRead: 199728
loops: 32, zonZeroPolls: 7
messagesRead: 199728
loops: 33, zonZeroPolls: 7
messagesRead: 199728
loops: 34, zonZeroPolls: 8
messagesRead: 216372
loops: 35, zonZeroPolls: 8
messagesRead: 216372
loops: 36, zonZeroPolls: 8
messagesRead: 216372
loops: 37, zonZeroPolls: 8
messagesRead: 216372
loops: 38, zonZeroPolls: 9
messagesRead: 249660
loops: 39, zonZeroPolls: 9
messagesRead: 249660
loops: 40, zonZeroPolls: 9
messagesRead: 249660
loops: 41, zonZeroPolls: 9
messagesRead: 249660
loops: 42, zonZeroPolls: 10
messagesRead: 266304
loops: 43, zonZeroPolls: 10
messagesRead: 266304
loops: 44, zonZeroPolls: 10
messagesRead: 266304
loops: 45, zonZeroPolls: 11
messagesRead: 282948
loops: 46, zonZeroPolls: 11
messagesRead: 282948
loops: 47, zonZeroPolls: 12
messagesRead: 299592
loops: 48, zonZeroPolls: 13
messagesRead: 316236
loops: 49, zonZeroPolls: 13
messagesRead: 316236
loops: 50, zonZeroPolls: 13
messagesRead: 316236
loops: 51, zonZeroPolls: 14
messagesRead: 332880
loops: 52, zonZeroPolls: 14
messagesRead: 332880
loops: 53, zonZeroPolls: 15
messagesRead: 366168
loops: 54, zonZeroPolls: 15
messagesRead: 366168
loops: 55, zonZeroPolls: 15
messagesRead: 366168
loops: 56, zonZeroPolls: 15
messagesRead: 366168
loops: 57, zonZeroPolls: 16
messagesRead: 382812
loops: 58, zonZeroPolls: 16
messagesRead: 382812
loops: 59, zonZeroPolls: 17
messagesRead: 399456
loops: 60, zonZeroPolls: 18
messagesRead: 416100
loops: 61, zonZeroPolls: 18
messagesRead: 416100
loops: 62, zonZeroPolls: 18
messagesRead: 416100
loops: 63, zonZeroPolls: 18
messagesRead: 416100
loops: 64, zonZeroPolls: 19
messagesRead: 432744
loops: 65, zonZeroPolls: 19
messagesRead: 432744
loops: 66, zonZeroPolls: 20
messagesRead: 449388
loops: 67, zonZeroPolls: 20
messagesRead: 449388
loops: 68, zonZeroPolls: 21
messagesRead: 466032
loops: 69, zonZeroPolls: 21
messagesRead: 466032
loops: 70, zonZeroPolls: 21
messagesRead: 466032
loops: 71, zonZeroPolls: 21
messagesRead: 466032
loops: 72, zonZeroPolls: 22
messagesRead: 499320
loops: 73, zonZeroPolls: 22
messagesRead: 499320
loops: 74, zonZeroPolls: 22
messagesRead: 499320
loops: 75, zonZeroPolls: 23
messagesRead: 515964
loops: 76, zonZeroPolls: 23
messagesRead: 515964
loops: 77, zonZeroPolls: 23
messagesRead: 515964
loops: 78, zonZeroPolls: 24
messagesRead: 532608
loops: 79, zonZeroPolls: 24
messagesRead: 532608
loops: 80, zonZeroPolls: 25
messagesRead: 549252
loops: 81, zonZeroPolls: 25
messagesRead: 549252
loops: 82, zonZeroPolls: 26
messagesRead: 565896
loops: 83, zonZeroPolls: 27
messagesRead: 582540
loops: 84, zonZeroPolls: 27
messagesRead: 582540
loops: 85, zonZeroPolls: 28
messagesRead: 599184
loops: 86, zonZeroPolls: 28
messagesRead: 599184
loops: 87, zonZeroPolls: 28
messagesRead: 599184
loops: 88, zonZeroPolls: 28
messagesRead: 599184
loops: 89, zonZeroPolls: 29
messagesRead: 615828
loops: 90, zonZeroPolls: 30
messagesRead: 632472
loops: 91, zonZeroPolls: 30
messagesRead: 632472
loops: 92, zonZeroPolls: 30
messagesRead: 632472
loops: 93, zonZeroPolls: 31
messagesRead: 649116
loops: 94, zonZeroPolls: 31
messagesRead: 649116
loops: 95, zonZeroPolls: 31
messagesRead: 649116
loops: 96, zonZeroPolls: 32
messagesRead: 682404
loops: 97, zonZeroPolls: 32
messagesRead: 682404
loops: 98, zonZeroPolls: 32
messagesRead: 682404
loops: 99, zonZeroPolls: 33
messagesRead: 699048
loops: 100, zonZeroPolls: 33
messagesRead: 699048
loops: 101, zonZeroPolls: 33
messagesRead: 699048
loops: 102, zonZeroPolls: 34
messagesRead: 715692
loops: 103, zonZeroPolls: 35
messagesRead: 732336
loops: 104, zonZeroPolls: 35
messagesRead: 732336
loops: 105, zonZeroPolls: 35
messagesRead: 732336
loops: 106, zonZeroPolls: 36
messagesRead: 748980
loops: 107, zonZeroPolls: 36
messagesRead: 748980
loops: 108, zonZeroPolls: 36
messagesRead: 748980
loops: 109, zonZeroPolls: 37
messagesRead: 765624
loops: 110, zonZeroPolls: 38
messagesRead: 782268
loops: 111, zonZeroPolls: 38
messagesRead: 782268
loops: 112, zonZeroPolls: 39
messagesRead: 798912
loops: 113, zonZeroPolls: 39
messagesRead: 798912
loops: 114, zonZeroPolls: 39
messagesRead: 798912
loops: 115, zonZeroPolls: 40
messagesRead: 815556
loops: 116, zonZeroPolls: 41
messagesRead: 832200
loops: 117, zonZeroPolls: 41
messagesRead: 832200
loops: 118, zonZeroPolls: 41
messagesRead: 832200
loops: 119, zonZeroPolls: 42
messagesRead: 848844
loops: 120, zonZeroPolls: 42
messagesRead: 848844
loops: 121, zonZeroPolls: 42
messagesRead: 848844
loops: 122, zonZeroPolls: 42
messagesRead: 848844
loops: 123, zonZeroPolls: 43
messagesRead: 865488
loops: 124, zonZeroPolls: 43
messagesRead: 865488
loops: 125, zonZeroPolls: 44
messagesRead: 882132
loops: 126, zonZeroPolls: 44
messagesRead: 882132
loops: 127, zonZeroPolls: 45
messagesRead: 898776
loops: 128, zonZeroPolls: 45
messagesRead: 898776
loops: 129, zonZeroPolls: 45
messagesRead: 898776
loops: 130, zonZeroPolls: 46
messagesRead: 915420
loops: 131, zonZeroPolls: 47
messagesRead: 915874
loops: 132, zonZeroPolls: 47
messagesRead: 915874
loops: 133, zonZeroPolls: 48
messagesRead: 932518
loops: 134, zonZeroPolls: 49
messagesRead: 949162
loops: 135, zonZeroPolls: 49
messagesRead: 949162
loops: 136, zonZeroPolls: 49
messagesRead: 949162
loops: 137, zonZeroPolls: 49
messagesRead: 949162
loops: 138, zonZeroPolls: 49
messagesRead: 949162
loops: 139, zonZeroPolls: 49
messagesRead: 949162
loops: 140, zonZeroPolls: 50
messagesRead: 982450
loops: 141, zonZeroPolls: 51
messagesRead: 982902
loops: 142, zonZeroPolls: 51
messagesRead: 982902
loops: 143, zonZeroPolls: 51
messagesRead: 982902
loops: 144, zonZeroPolls: 51
messagesRead: 982902
loops: 145, zonZeroPolls: 51
messagesRead: 982902
loops: 146, zonZeroPolls: 51
messagesRead: 982902
loops: 147, zonZeroPolls: 52
messagesRead: 999546
Elapsed time: 12.921 seconds
all finished! messagesRead: 1000000, bytesRead: 100000000
loops: 148, zonZeroPolls: 53
2015-08-30 23:36:53:269, 2015-08-30 23:37:06:191, 95.3674, 7.3802, 1000000, 77387.4013
```

Commit:
Various cleanup per review: - perf services pipe output to separate log file; removed default values from ConsumerPerformancService; further consolidated the test_producer_throughput methods.

granders commented:
Update - I rebased off of the consumer performance fix (https://github.com/apache/kafka/pull/180). We're still seeing the timeout in the consume loop; @hachikuji is looking into it

Commit:
Increase timeout on new-consumer so that ConsumerPerformance.scala can be run on local VMs.

hachikuji commented:
I've been able to reproduce the timeout locally on a fresh cluster. It appears to be caused by the delay required to create the __consumer_offsets topic. The consumer metadata request currently depends on this topic in order to discover the coordinator of a group. Once it's created, however, I haven't been able to reproduce the timeout.

ewencp commented:
@hachikuji So what's the proposed solution here? I see a couple of options:
1. Increase the timeout here/use a longer timeout for the first poll() so the broker has extra time to create the topic. It's possible we want to do this anyway because it may be too low for a join group cycle to complete with default settings if you have multiple consumers in the group. However, this is just a hack around the real issue, which is that consumer metadata lookup can block for a long time waiting for topic creation. Any code that might interact with a fresh cluster would need similar tweaking.
2. Create the offsets topic proactively. It looks like right now it might only be created upon consumer metadata request. Is there a good reason we can't create it earlier? Then system tests like these could just wait until it's been created before trying to proceed (we can monitor the logs for the relevant output). It also probably provides a better out-of-the-box experience for users -- it's kind of weird to potentially have your first interaction with a cluster as a consumer potentially fail... On the other hand, this fix only works well for system tests because they would _know_ to monitor for that event; new users just running CLI commands wouldn't know about that, so it could still be a problem if the topic takes too long to create.

hachikuji commented:
@ewencp I think @granders said he thought there were cases he was seeing timeouts where this wouldn't have been an issue, so I was going to wait to see what he could find. The option I discussed with him was to use the poll(0) trick to ensure that the assignment is active before beginning the consume loop. The better solution would probably be to combine waitUntil and the rebalance callback to get the same effect.

ewencp commented:
@hachikuji Yeah, I wanted to make sure we have a good solution for this because it's actually a pretty big usability issue for the new consumer, in my opinion. The poll(0) trick works, but a) isn't great since poll(0) shouldn't actually block anyway (if we fix timeouts on most things, this won't really work anymore) and b) requires every app that actually writes real tests to do the same thing because in all likelihood, if they perform an integration or system test, they'll be working with a fresh cluster.

hachikuji commented:
@ewencp The better solution is to use the rebalance callback, but it's pretty awkward in code:

``` java
final AtomicBoolean isAssigned = new AtomicBoolean(false);
consumer.subscribe(topics, new ConsumerRebalanceListener() {
  void onAssign(partitions) {
    isAssigned.set(true);
  }
  void onRevoke(partitions) {
    isAssigned.set(false);
  }
}
while (!isAssigned.get())
  consumer.poll(100);

consumer.seekToBeginning()
```

Ouch. If this is a common enough need, maybe we should consider exposing something like `ensureAssignment` which has the same effect.

Commit:
Updated ConsumerPerformance.scala so that when new consumer is used, the initial step of fetching metadata, groupjoining etc, happens outside of the benchmark loop.

granders commented:
@ewencp @hachikuji I updated `ConsumerPerformance.scala` to use the better solution, and this seems to have solved the issue. Also, added logging.

ewencp commented:
@granders Looks good now. I'm getting real numbers now, although some of them might need some further investigation.

@hachikuji The code is, in fact, pretty verbose. Shall we file a JIRA to figure out if we should do anything API-wise to make this less ridiculous (especially since if you want to use the rebalance callback for anything else, it'll get significantly messier...)?

@gwenshap Probably you're up for review here since you've handled all the previous system test stuff?

gwenshap commented on core/src/main/scala/kafka/tools/ConsumerPerformance.scala:
```diff
@@ -93,18 +97,40 @@ object ConsumerPerformance {
+      }})
+    val joinStart = System.currentTimeMillis()
+    while (!isAssigned.get()) {
+      if (System.currentTimeMillis() - joinStart >= joinTimeout) {
```
Since we are benchmarking, does it make sense to add the join time to the output? We'll want to know if there are significant changes there.

granders replied:
@gwenshap 
I created a new JIRA to track this: https://issues.apache.org/jira/browse/KAFKA-2525

gwenshap commented on tests/kafkatest/services/templates/tools_log4j.properties:
```diff
@@ -14,7 +14,7 @@
 
 # Define the root logger with appender file
-log4j.rootLogger = INFO, FILE
+log4j.rootLogger = {{ log_level|default("INFO") }}, FILE
```
Nice. I did not know you could do that.

gwenshap commented:
LGTM. Just one small suggestion above.

Commit:
Re-added changed introduced by https://github.com/apache/kafka/pull/158

b-goyal commented:
@ewencp 
The changes here are smaller than they look - mostly refactoring/cleanup.
- ConsumerPerformanceService: added new_consumer flag, and exposed more command-line settings
- benchmark.py: refactored to use `@parametrize` and `@matrix` - this reduced some amount of repeated code
- benchmark.py: added consumer performance tests with new consumer (using `@parametrize`)
- benchmark.py: added more detailed test descriptions
- performance.py: broke into separate files
